content/blog/redux-saga-style-guide.md/index.md

```diff
@@ -0,0 +1,350 @@
+
+- https://redux.js.org/style-guide/style-guide#allow-many-reducers-to-respond-to-the-same-action
+
+My hot take is that I think there should be a 1:1 mapping between actions and
+reducers. Reducers should own the actions (via `createSlice`) and only under
+rare exceptions should a reducer listen to outside actions.
+
+Employing this method `redux` becomes a very thin layer of setters to hold data
+and tell `react` when state has changed. To me, the real value of redux is:
+
+- One structured way to update state
+- A single source of truth that our UI _reacts_ to
+- With `react-redux` synchronizing UI with state is handled automatically
+- Reducers are pure functions that are easy to test
+
+I'll agree that putting more logic into reducers makes it easier to test, but
+I'll also argue that it makes it more difficult to maintain and understand.
+
+### Avoid dispatching many actions sequentially
+
+- https://redux.js.org/style-guide/style-guide#avoid-dispatching-many-actions-sequentially
+
+I want a list of steps that demonstrate how redux is being updated. What I don't
+want to do is grep all the reducers for an action type to see how my state is
+being updated. This is especially annoying when employing a modular,
+feature-based folder structure. We have replaced a single function that
+centralized business logic into a composition of functions that are scattered
+throughout the codebase. The logic is broken up which makes the flow of what is
+happening harder to understand. What compounds this even worse, with libraries
+like `redux-saga`, sagas can also listen for those actions and activate even
+more side-effects. Generally speaking, I try to only let sagas listen for events
+(react-side), not my setters.
+
+## Saga style-guide
+
+Take the `redux` style-guide, remove the ones listed above, and add these for my
+unofficial `redux-saga` style-guide.
+
+### Reducers as setters
+
+Redux is an object that should be thought of like a database. Reducer slices are
+database tables. We should reduce boilerplate with slice helpers
+([robodux](https://github.com/neurosnap/robodux#slice-helpers) or
+[slice-helpers](https://github.com/neurosnap/slice-helpers)) by leveraging new
+_officially_ sanctioned helpers like `createSlice`.
+
+We don't even need to test our reducers anymore because these libraries already
+did that for us.
+
+This makes reducers predictable, isn't that one of the taglines for redux? A
+`predictable` state container? Reducers are simplified, and slice helpers cover
+90% of our use cases, because we are treating them like database tables.
+
+### Build indexes for your db tables
+
+Need to have a sorted list of entities? Need to group a subset of entities?
+Create a reducer that acts like an index.
+
+Yes, it feels like we are rebuilding a database, but it's not that much work and
+the manual process allows for performance tweaking which is required when build
+a large application.
+
+TODO: add code sample
```

Comment:
Agreed.  The current rule I operate under is this:  If it can be derived state then it should.  However, this is not always the case.  If we want to get all the people associated with a Role and that information is not on the person or the role and we have to make a separate request to get that information, then it must be an index.  `rolePeopleIds` and `peopleRoleIds` are those cases in comply.

Reply:
Ah, that makes sense.

Since `rolePeopleIds` and `peopleRoleIds` are separate endpoints, they should be treated as separately in Redux too. That makes sure they both get properly loaded and updated.

Reply:
I don't think it necessarily has to be this way, I think there might be room for API improvements, but the fact remains, there are situations where we need to build the index as a reducer.